x/valset/keeper/keeper.go

```diff
@@ -729,3 +736,96 @@ func (k Keeper) SaveModifiedSnapshot(ctx context.Context, snapshot *types.Snapsh
 	snapStore := k.snapshotStore(ctx)
 	return keeperutil.Save(snapStore, k.cdc, keeperutil.Uint64ToByte(snapshot.GetId()), snapshot)
 }
+
+// Set the current requirements for pigeon. This will also clear any scheduled
+// requirements.
+func (k Keeper) SetPigeonRequirements(ctx context.Context, req *types.PigeonRequirements) error {
+	if req == nil {
+		return nil
+	}
+
+	pigeonStore := k.pigeonStore(ctx)
+
+	pigeonReq, err := k.PigeonRequirements(ctx)
+	if err != nil {
+		return err
+	}
+
+	if semver.Compare(req.MinVersion, pigeonReq.MinVersion) < 0 {
+		liblog.FromSDKLogger(k.Logger(ctx)).
+			WithFields(
+				"req-min-version", req.MinVersion,
+				"cur-min-version", pigeonReq.MinVersion,
+			).Error("new version cannot be lower than current")
+		return errors.New("new version cannot be lower than current")
+	}
+
+	err = keeperutil.Save(pigeonStore, k.cdc, pigeonRequirementsKey, req)
+	if err != nil {
+		return err
+	}
+
+	// Delete scheduled requirements
+	pigeonStore.Delete(pigeonScheduledRequirementsKey)
```

Comment:
What happens in case the save goes through but the delete fails? Will it be re-attempted and then continously fail since the version isn't higher anymore?

If so, use a branched context and only commit in case call operations go through.

Reply:
I added the CacheContext to both, but I'm not 100% sure it's right. The `Delete` method does not return an error, so it might still be possible to fall into the same situation?

We could be more defensive about this by having `SetPigeonRequirements` return a specific error for version mismatch. This could be detected in `BeginBlock` and we clear the value explicitly from there. What do you think?

Reply:
Hmm. Maybe the more straight-forward solution would be to alter the `Delete` method to return an error. I wonder that's not happening. Is it coming straight from the Cosmos framework or is there some wrapper magic happening around it?

Reply:
It is coming straight from the SDK. We get a `KVStore` that embeds a [BasicKVStore](https://github.com/cosmos/cosmos-sdk/blob/main/store/types/store.go#L233) that does not return errors on any operation. The `Save` is [our own wrapper](https://github.com/palomachain/paloma/blob/master/util/keeper/save_load.go#L10) and just returns marshalling errors.

Reply:
It seems the concrete store implementations [may panic](https://github.com/cosmos/cosmos-sdk/blob/main/store/iavl/store.go#L198) instead of returning an error. In that case, the CacheContext and commit after Delete should work as all-or-nothing and should be fine. There may still be the case where there's some fs corruption affecting one store and not the other, but the node will have bigger problems then.